modules/ml/src/main/java/org/apache/ignite/ml/dataset/UpstreamTransformer.java

```diff
@@ -28,15 +27,8 @@
  * @param <V> Type of values in the upstream.
  */
 @FunctionalInterface
-public interface UpstreamTransformer<K, V> extends Serializable {
-    /**
-     * Perform transformation of upstream.
-     *
-     * @param rnd Random numbers generator.
-     * @param upstream Upstream.
-     * @return Transformed upstream.
-     */
-    // TODO: IGNITE-10296: Inject capabilities of randomization through learning environment.
-    // TODO: IGNITE-10297: Investigate possibility of API change.
-    public Stream<UpstreamEntry<K, V>> transform(Random rnd, Stream<UpstreamEntry<K, V>> upstream);
+public interface UpstreamTransformer<K, V> extends IgniteFunction<Stream<UpstreamEntry<K, V>>, Stream<UpstreamEntry<K, V>>> {
+    default UpstreamTransformer<K, V> then(UpstreamTransformer<K, V> other) {
```

Comment:
Thanks for javadoc and 'public', as for `then` the point was to
1. save interoperability of `UpstreamTransformer` with plain functions (hence inheritance from `IgniteFunction`);
 2. make combiner which is like `andThen`, but under which `UpstreamTransformers` are closed (`upstreamTransformer.then(upstreamTransformer)` produces `UpstreamTransformer`, not just a `Function`).
If we drop interoperability thing, then sure, right way to go would be `andThen`. Maybe we can drop it, because `UpstreamTransformer` will be, I guess, primarily used with other `UpstreamTransformer`s.

Reply:
I think we could eliminate inheritance from `IgniteFunction` here. I don't see any benefits of it, but two methods like `then` and `andThen` are very confusing.

Reply:
Yeah, already done that.